### PR TITLE
ZCL_ABAPGIT_REPO_ONLINE: Change methods fetch_remote and push to be able to retrieve and push objects also by commit hash

### PR DESCRIPTION
**Pull request by mariusraht2**
Related to #3040

Objects can now be retrieved and pushed/staged by commit hash and not only the branch hash, whereas the commit hash is only set if the user doesn't use the branch hash as top commit.

The method fetch_remote is being changed to be able to retrieve remote objects if a repo commit has been set. If only the branch name is supplied we have to get the remote objects from the top (root/branch) commit.

The method push is being changed to be able to push local objects if a repo commit has been set. If only the branch hash is supplied we have to push the objects with binding to the top (root/branch) commit.

Additionally this is also needed to test the changes in this PR: In case the branch is being switched the commit hash has to be reset.

### CODE REVIEW

File: src/zcl_abapgit_repo_online.clas.abap [hunk 1]
```diff
@@ -406,12 +411,18 @@ CLASS zcl_abapgit_repo_online IMPLEMENTATION.
 
     handle_stage_ignore( io_stage ).
 
+    IF get_sha1( ) IS INITIAL.
+      lv_parent = get_sha1_remote( ).
+    ELSE.
+      lv_parent = get_sha1( ).
```
Comment (larshp) [inline src/zcl_abapgit_repo_online.clas.abap:417]: what is the difference between `get_sha1_remote` and `get_sha1`, and how can the developer tell which to use when?

**Review by larshp — DISMISSED** [review]

File: src/zcl_abapgit_repo_online.clas.abap (hunk 1 above)
Comment (mariusraht2) [reply src/zcl_abapgit_repo_online.clas.abap:443]: s. PR #4082

**Comment by mariusraht2** [comment]
@mariusraht2 TODO: Update source code based on PR #4086

**Comment by larshp** [comment]
this is good, now I can follow what is happening 👍

**Pull request merged**